File: src/org/hsluv/HSLuvColorSpace.java
package org.hsluv;

import org.joml.Vector2d;

public class HSLuvColorSpace {

    private static final double[] xyz2rgb = new double[]{
            3.240969941904521, -1.537383177570093, -0.498610760293,
            -0.96924363628087, 1.87596750150772, 0.041555057407175,
            0.055630079696993, -0.20397695888897, 1.056971514242878,
    };

    private static final double[] rgb2xyz = new double[]{
            0.41239079926595, 0.35758433938387, 0.18048078840183,
            0.21263900587151, 0.71516867876775, 0.072192315360733,
            0.019330818715591, 0.11919477979462, 0.95053215224966,
    };

    private static final double refY = 1.0;

    private static final double refU = 0.19783000664283;
    private static final double refV = 0.46831999493879;

    private static final double kappa = 903.2962962;
    private static final double epsilon = 0.0088564516;

    private static Vector2d[] getBounds(double L, int limit) {

        Vector2d[] result = new Vector2d[limit];

        double sub1 = Math.pow(L + 16.0, 3.0) / 1560896.0;
        double sub2 = sub1 > epsilon ? sub1 : L / kappa;

        for (int c = 0, index = 0; c < 3; c++) {

            double m1 = xyz2rgb[c * 3];
            double m2 = xyz2rgb[c * 3 + 1];
            double m3 = xyz2rgb[c * 3 + 2];

            for (int t = 0; t < 2; t++) {
                double top1 = (284517 * m1 - 94839 * m3) * sub2;
                double top2 = (838422 * m3 + 769860 * m2 + 731718 * m1) * L * sub2 - 769860 * t * L;
                double bottom = (632260 * m3 - 126452 * m2) * sub2 + 126452 * t;
                result[index++] = new Vector2d(top1 / bottom, top2 / bottom);
                if (index >= limit) return result;
            }
        }

        return result;

    }

    private static double intersectLineLine(Vector2d lineA, Vector2d lineB) {
        return (lineA.y - lineB.y) / (lineB.x - lineA.x);
    }

    private static double distanceFromPole(Vector2d point) {
        return point.length();
    }

    private static double lengthOfRayUntilIntersect(double theta, Vector2d line) {
        return line.y / (Math.sin(theta) - line.x * Math.cos(theta));
    }

    private static double maxSafeChromaForL(double L) {
        Vector2d[] bounds = getBounds(L, 2);
        double min = Double.MAX_VALUE;

        for (int i = 0; i < 2; i++) {// why only the first two bounds? because they must belong to L

            Vector2d line0 = bounds[i];
            double m1 = line0.x;
            double b1 = line0.y;

            Vector2d line = new Vector2d(m1, b1);

            double x = intersectLineLine(line, new Vector2d(-1.0 / m1, 0.0));
            double length = distanceFromPole(new Vector2d(x, b1 + x * m1));

            min = Math.min(min, length);

        }

        return min;
    }

    private static double maxChromaForLH(double L, double H) {
        double hRadians = H / 360 * Math.PI * 2;

        Vector2d[] bounds = getBounds(L,6);
        double min = Double.MAX_VALUE;

        for (Vector2d bound : bounds) {
            double length = lengthOfRayUntilIntersect(hRadians, bound);
            if (length >= 0.0) {
                min = Math.min(min, length);
            }
        }

        return min;
    }

    private static double dotProduct(double[] a, int aOffset, double[] b) {
        double sum = 0;
        for (int i = 0; i < b.length; ++i) {
            sum += a[i + aOffset] * b[i];
        }
        return sum;
    }

    private static double fromLinear(double c) {
        if (c <= 0.0031308) {
            return 12.92 * c;
        } else {
            return 1.055 * Math.pow(c, 1 / 2.4) - 0.055;
        }
    }

    private static double toLinear(double c) {
        if (c > 0.04045) {
            return Math.pow((c + 0.055) / (1 + 0.055), 2.4);
        } else {
            return c / 12.92;
        }
    }

    /*private static int[] rgbPrepare(double[] tuple) {

        int[] results = new int[tuple.length];

        for (int i = 0; i < tuple.length; ++i) {
            double chan = tuple[i];
            double rounded = round(chan, 3);

            if (rounded < -0.0001 || rounded > 1.0001) {
                throw new IllegalArgumentException("Illegal rgb value: " + rounded);
            }

            results[i] = (int) Math.round(rounded * 255);
        }

        return results;
    }*/

    public static double[] xyzToRgb(double[] tuple) {
        return new double[]{
                fromLinear(dotProduct(xyz2rgb, 0, tuple)),
                fromLinear(dotProduct(xyz2rgb, 3, tuple)),
                fromLinear(dotProduct(xyz2rgb, 6, tuple)),
        };
    }

    public static double[] rgbToXyz(double[] tuple) {
        double[] linearRGB = new double[]{
                toLinear(tuple[0]),
                toLinear(tuple[1]),
                toLinear(tuple[2]),
        };
        return new double[]{
                dotProduct(rgb2xyz, 0, linearRGB),
                dotProduct(rgb2xyz, 3, linearRGB),
                dotProduct(rgb2xyz, 6, linearRGB),
        };
    }

    private static double yToL(double Y) {
        if (Y <= epsilon) {
            return (Y / refY) * kappa;
        } else {
            return 116.0 * Math.pow(Y / refY, 1.0 / 3.0) - 16.0;
        }
    }

    private static double lToY(double L) {
        if (L <= 8.0) {
            return refY * L / kappa;
        } else {
            return refY * Math.pow((L + 16.0) / 116.0, 3.0);
        }
    }

    public static double[] xyzToLuv(double[] tuple) {

        double X = tuple[0];
        double Y = tuple[1];
        double Z = tuple[2];

        double varU = (4.0 * X) / (X + (15.0 * Y) + (3.0 * Z));
        double varV = (9.0 * Y) / (X + (15.0 * Y) + (3.0 * Z));

        double L = yToL(Y);

        if (L == 0) {
            return new double[]{0, 0, 0};
        }

        double U = 13.0 * L * (varU - refU);
        double V = 13.0 * L * (varV - refV);

        return new double[]{L, U, V};

    }

    public static double[] luvToXyz(double[] tuple) {
        double L = tuple[0];
        double U = tuple[1];
        double V = tuple[2];

        if (L == 0) {
            return new double[]{0, 0, 0};
        }

        double varU = U / (13.0 * L) + refU;
        double varV = V / (13.0 * L) + refV;

        double Y = lToY(L);
        double X = 0 - (9.0 * Y * varU) / ((varU - 4.0) * varV - varU * varV);
        double Z = (9.0 * Y - (15.0 * varV * Y) - (varV * X)) / (3.0 * varV);

        return new double[]{X, Y, Z};
    }

    public static double[] luvToLch(double[] tuple) {

        double L = tuple[0];
        double U = tuple[1];
        double V = tuple[2];

        double C = Math.sqrt(U * U + V * V);
        double H;

        if (C < 0.00000001) {

            H = 0;

        } else {

            H = StrictMath.toDegrees(Math.atan2(V, U));

            if (H < 0) {
                H += 360;
            }

        }

        return new double[]{L, C, H};

    }

    public static double[] lchToLuv(double[] tuple) {

        double L = tuple[0];
        double C = tuple[1];
        double H = tuple[2];

        double hRadians = StrictMath.toRadians(H);
        double U = Math.cos(hRadians) * C;
        double V = Math.sin(hRadians) * C;

        return new double[]{L, U, V};

    }

    public static double[] hsluvToLch(double[] tuple) {

        double H = tuple[0];
        double S = tuple[1];
        double L = tuple[2];

        if (L > 99.9999999) {
            return new double[]{100, 0, H};
        }

        if (L < 0.00000001) {
            return new double[]{0, 0, H};
        }

        double max = maxChromaForLH(L, H);
        double C = max * 0.01 * S;

        return new double[]{L, C, H};

    }

    public static double[] lchToHsluv(double[] tuple) {
        double L = tuple[0];
        double C = tuple[1];
        double H = tuple[2];

        if (L > 99.9999999) {
            return new double[]{H, 0, 100};
        }

        if (L < 0.00000001) {
            return new double[]{H, 0, 0};
        }

        double max = maxChromaForLH(L, H);
        double S = C / max * 100;

        return new double[]{H, S, L};
    }

    public static double[] hpluvToLch(double[] tuple) {
        double H = tuple[0];
        double S = tuple[1];
        double L = tuple[2];

        if (L > 99.9999999) {
            return new double[]{100, 0, H};
        }

        if (L < 0.00000001) {
            return new double[]{0, 0, H};
        }

        double max = maxSafeChromaForL(L);
        double C = max / 100 * S;

        return new double[]{L, C, H};
    }

    public static double[] lchToHpluv(double[] tuple) {
        double L = tuple[0];
        double C = tuple[1];
        double H = tuple[2];

        if (L > 99.9999999) {
            return new double[]{H, 0, 100};
        }

        if (L < 0.00000001) {
            return new double[]{H, 0, 0};
        }

        double max = maxSafeChromaForL(L);
        double S = C / max * 100;

        return new double[]{H, S, L};
    }

    public static double[] lchToRgb(double[] tuple) {
        return xyzToRgb(luvToXyz(lchToLuv(tuple)));
    }

    public static double[] rgbToLch(double[] tuple) {
        return luvToLch(xyzToLuv(rgbToXyz(tuple)));
    }

    public static double[] hsluvToRgb(double[] tuple) {
        return lchToRgb(hsluvToLch(tuple));
    }

    public static double[] rgbToHsluv(double[] tuple) {
        return lchToHsluv(rgbToLch(tuple));
    }

    public static double[] hpluvToRgb(double[] tuple) {
        return lchToRgb(hpluvToLch(tuple));
    }

    public static double[] rgbToHpluv(double[] tuple) {
        return lchToHpluv(rgbToLch(tuple));
    }

}

File: src/org/hsluv/HSLuvColorSpace.kt
package org.hsluv

import me.anno.utils.hpc.threadLocal
import me.anno.utils.types.Floats.toDegrees
import me.anno.utils.types.Floats.toRadians
import org.joml.Vector2d
import org.joml.Vector3d
import org.joml.Vector3f
import kotlin.math.PI
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.hypot
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * HSL, XYZ and HSLuv conversion from https://github.com/hsluv/hsluv-java
 * converted to Kotlin & made to use JOML vectors
 * */
@Suppress("MemberVisibilityCanBePrivate", "unused")
object HSLuvColorSpace {

    private val xyz2rgb = doubleArrayOf(
        3.240969941904521, -1.537383177570093, -0.498610760293,
        -0.96924363628087, 1.87596750150772, 0.041555057407175,
        0.055630079696993, -0.20397695888897, 1.056971514242878,
    )

    private val xyz2rgb2 = floatArrayOf(
        3.24097f, -1.5373832f, -0.49861076f,
        -0.96924365f, 1.8759675f, 0.04155506f,
        0.05563008f, -0.20397696f, 1.0569715f,
    )

    private val rgb2xyz = doubleArrayOf(
        0.41239079926595, 0.35758433938387, 0.18048078840183,
        0.21263900587151, 0.71516867876775, 0.072192315360733,
        0.019330818715591, 0.11919477979462, 0.95053215224966,
    )

    private const val refY = 1.0

    private const val refU = 0.19783000664283
    private const val refV = 0.46831999493879

    private const val kappa = 903.2962962
    private const val epsilon = 0.0088564516

    private val threadLocalBounds = threadLocal { Array(6) { Vector2d() } }

    private fun getBounds(l: Double, limit: Int): Array<Vector2d> {

        val result = threadLocalBounds.get()

        val sub1 = (l + 16.0).pow(3.0) / 1560896.0
        val sub2 = if (sub1 > epsilon) sub1 else l / kappa

        var index = 0
        for (c in 0 until 3) {

            val m1 = xyz2rgb[c * 3]
            val m2 = xyz2rgb[c * 3 + 1]
            val m3 = xyz2rgb[c * 3 + 2]

            for (t in 0 until 2) {
                val top1 = (284517 * m1 - 94839 * m3) * sub2
                val top2 = (838422 * m3 + 769860 * m2 + 731718 * m1) * l * sub2 - 769860 * t * l
                val bottom = (632260 * m3 - 126452 * m2) * sub2 + 126452 * t
                result[index++].set(top1 / bottom, top2 / bottom)
                if (index >= limit) return result
            }
        }

        return result
    }

    private fun intersectLineLine(lineAx: Double, lineAy: Double, lineBx: Double): Double {
        // lineB.y is always 0
        return lineAy / (lineBx - lineAx)
    }

    private fun lengthOfRayUntilIntersect(theta: Double, line: Vector2d): Double {
        return line.y / (sin(theta) - line.x * cos(theta))
    }

    private fun maxSafeChromaForL(l: Double): Double {

        val bounds = getBounds(l, 2)
        var min = Double.MAX_VALUE

        for (i in 0 until 2) {// why only the first two bounds? because they must belong to l

            val line0 = bounds[i]
            val m1 = line0.x
            val b1 = line0.y

            val x = intersectLineLine(m1, b1, -1.0 / m1)
            val length = hypot(x, b1 + x * m1)

            min = min(min, length)
        }

        return min
    }

    private fun maxChromaForLH(l: Double, h: Double): Double {
        val hRadians = h / 360 * PI * 2

        val bounds = getBounds(l, 6)
        var min = Double.MAX_VALUE

        for (i in 0 until 6) {
            val bound = bounds[i]
            val length = lengthOfRayUntilIntersect(hRadians, bound)
            if (length >= 0.0) {
                min = min(min, length)
            }
        }

        return min
    }

    private fun dotProduct(a: DoubleArray, aOffset: Int, b: Vector3d): Double {
        return b.dot(a[aOffset], a[aOffset + 1], a[aOffset + 2])
    }

    private fun xyzToRGB(aOffset: Int, b: Vector3f): Float {
        return b.dot(xyz2rgb2[aOffset], xyz2rgb2[aOffset + 1], xyz2rgb2[aOffset + 2])
    }

    /**
     * converts a value from linear to sRGB
     * */
    fun toSRGB(c: Double): Double {
        return if (c <= 0.0031308) {
            12.92 * c
        } else {
            1.055 * c.pow(1.0 / 2.4) - 0.055
        }
    }

    /**
     * converts a value from linear to sRGB
     * */
    fun toSRGB(c: Float): Float {
        return toSRGB(c.toDouble()).toFloat()
    }

    /**
     * converts a value from sRGB to linear
     * */
    fun toLinear(c: Double): Double {
        return if (c > 0.04045) {
            ((c + 0.055) / (1.055)).pow(2.4)
        } else {
            c / 12.92
        }
    }

    /**
     * converts a value from sRGB to linear
     * */
    fun toLinear(c: Float): Float {
        return toLinear(c.toDouble()).toFloat()
    }

    fun xyzToRgb(src: Vector3d, dst: Vector3d = src): Vector3d {
        return dst.set(
            toSRGB(dotProduct(xyz2rgb, 0, src)),
            toSRGB(dotProduct(xyz2rgb, 3, src)),
            toSRGB(dotProduct(xyz2rgb, 6, src)),
        )
    }

    fun xyzToRgb(src: Vector3f, dst: Vector3f = src): Vector3f {
        return dst.set(
            toSRGB(xyzToRGB(0, src)),
            toSRGB(xyzToRGB(3, src)),
            toSRGB(xyzToRGB(6, src)),
        )
    }

    fun rgbToXyz(src: Vector3d, dst: Vector3d = src): Vector3d {
        dst.set(
            toLinear(src.x),
            toLinear(src.y),
            toLinear(src.z),
        )
        return dst.set(
            dotProduct(rgb2xyz, 0, dst),
            dotProduct(rgb2xyz, 3, dst),
            dotProduct(rgb2xyz, 6, dst),
        )
    }

    private fun yToL(y: Double) = if (y <= epsilon) {
        (y / refY) * kappa
    } else {
        116.0 * (y / refY).pow(1.0 / 3.0) - 16.0
    }

    private fun lToY(l: Double) = if (l <= 8.0) {
        refY * l / kappa
    } else {
        refY * ((l + 16.0) / 116.0).pow(3.0)
    }

    fun xyzToLuv(src: Vector3d, dst: Vector3d = src): Vector3d {

        val y = src.y

        val l = yToL(y)
        if (l == 0.0) return dst.set(0.0)

        val x = src.x
        val z = src.z

        val y15 = y * 15.0
        val z3 = z * 3.0
        val norm = 1.0 / (x + y15 + z3)
        val varU = 4.0 * x * norm
        val varV = 9.0 * y * norm

        val l13 = l * 13
        val u = l13 * (varU - refU)
        val v = l13 * (varV - refV)

        return dst.set(l, u, v)
    }

    fun luvToXyz(src: Vector3d, dst: Vector3d = src): Vector3d {

        val l = src.x
        if (l == 0.0) return dst.set(0.0)

        val u = src.y
        val v = src.z

        val l13 = 1.0 / (l * 13)
        val varU = u * l13 + refU
        val varV = v * l13 + refV

        val y = lToY(l)
        val x = 0 - (9.0 * y * varU) / ((varU - 4.0) * varV - varU * varV)
        val z = (9.0 * y - (15.0 * varV * y) - (varV * x)) / (3.0 * varV)

        return dst.set(x, y, z)
    }

    fun luvToLch(src: Vector3d, dst: Vector3d = src): Vector3d {

        val l = src.x
        val u = src.y
        val v = src.z

        val c = sqrt(u * u + v * v)
        var h: Double

        if (c < 0.00000001) {
            h = 0.0
        } else {
            h = atan2(v, u).toDegrees()
            if (h < 0.0) h += 360.0
        }

        return dst.set(l, c, h)
    }

    fun lchToLuv(src: Vector3d, dst: Vector3d = src): Vector3d {

        val l = src.x
        val c = src.y
        val h = src.z

        val hRadians = h.toRadians()
        val u = cos(hRadians) * c
        val v = sin(hRadians) * c

        return dst.set(l, u, v)
    }

    fun hsluvToLch(src: Vector3d, dst: Vector3d = src): Vector3d {

        val h = src.x
        val s = src.y
        val l = src.z

        if (l > 99.9999999) return dst.set(100.0, 0.0, h)
        if (l < 0.00000001) return dst.set(0.0, 0.0, h)

        val max = maxChromaForLH(l, h)
        val c = max * 0.01 * s

        return dst.set(l, c, h)
    }

    fun lchToHsluv(src: Vector3d, dst: Vector3d): Vector3d {
        val l = src.x
        val c = src.y
        val h = src.z

        if (l > 99.9999999) {
            return dst.set(h, 0.0, 100.0)
        }

        if (l < 0.00000001) {
            return dst.set(h, 0.0, 0.0)
        }

        val max = maxChromaForLH(l, h)
        val s = c / max * 100.0

        return dst.set(h, s, l)
    }

    fun hpluvToLch(src: Vector3d, dst: Vector3d = src): Vector3d {
        val h = src.x
        val s = src.y
        val l = src.z

        if (l > 99.9999999) return dst.set(100.0, 0.0, h)
        if (l < 0.00000001) return dst.set(0.0, 0.0, h)

        val max = maxSafeChromaForL(l)
        val c = max / 100 * s

        return dst.set(l, c, h)
    }

    fun lchToHpluv(src: Vector3d, dst: Vector3d = src): Vector3d {
        val l = src.x
        val c = src.y
        val h = src.z

        if (l > 99.9999999) return dst.set(h, 0.0, 100.0)
        if (l < 0.00000001) return dst.set(h, 0.0, 0.0)

        val max = maxSafeChromaForL(l)
        val s = c / max * 100.0

        return dst.set(h, s, l)
    }

    fun lchToRgb(src: Vector3d, dst: Vector3d = src) = xyzToRgb(luvToXyz(lchToLuv(src), dst), dst)
    fun rgbToLch(src: Vector3d, dst: Vector3d = src) = luvToLch(xyzToLuv(rgbToXyz(src, dst), dst), dst)
    fun hsluvToRgb(src: Vector3d, dst: Vector3d = src) = lchToRgb(hsluvToLch(src, dst), dst)
    fun rgbToHsluv(src: Vector3d, dst: Vector3d = src) = lchToHsluv(rgbToLch(src, dst), dst)
    fun hpluvToRgb(src: Vector3d, dst: Vector3d = src) = lchToRgb(hpluvToLch(src, dst), dst)
    fun rgbToHpluv(src: Vector3d, dst: Vector3d = src) = lchToHpluv(rgbToLch(src, dst), dst)
}